zerver/lib/narrow.py

```diff
@@ -919,6 +927,30 @@ def get_channel_from_narrow_access_unchecked(
     return None
 
 
+def update_narrow_terms_containing_with_operator(
+    realm: Realm, narrow: OptionalNarrowListT
+) -> OptionalNarrowListT:
+    if narrow is not None:
+        contains_with_operator = list(filter(lambda term: term["operator"] == "with", narrow))
+
+        if len(contains_with_operator) > 0:
+            with_term = contains_with_operator[0]
+            message = get_recipient_topic_by_narrow_operand_access_unchecked(
+                realm, int(with_term["operand"])
+            )
+            if message is not None and message.is_stream_message():
+                for term in narrow:
+                    if term["operator"] == "channel":
+                        stream_id = message.recipient.type_id
+                        term["operand"] = get_stream_by_id_in_realm(stream_id, realm).name
```

Comment:
Having to do this lookup is pretty ugly; but I guess the fix for that will require moving to generally sending stream IDs rather than names over the wire. Maybe worth adding a `TODO` comment that ideally we'd not have to translate back to a stream name here.

Reply:
The API does accept stream/channel IDs:
https://zulip.com/api/construct-narrow#stream-and-user-ids

So it should be possible to write this code to stay in terms of channel IDs too.

Reply:
added the TODO

Reply:
Updated to rather use channel id